src/Intervals.php

```diff
@@ -100,11 +106,107 @@ public static function haveIntersections(ConstraintInterface $a, ConstraintInter
             return true;
         }
 
+        if ($a instanceof MatchNoneConstraint || $b instanceof MatchNoneConstraint) {
+            return false;
+        }
+
         $intersectionIntervals = self::generateIntervals(new MultiConstraint(array($a, $b), true), true);
 
         return \count($intersectionIntervals['numeric']) > 0 || \count($intersectionIntervals['branches']) > 0;
     }
 
+    /**
+     * Attempts to optimize a MultiConstraint
+     *
+     * When merging MultiConstraints together they can get very large, this will
+     * compact it by looking at the real intervals covered by all the constraints
+     * and then creates a new constraint containing only the smallest amount of rules
+     * to match the same intervals.
+     *
+     * @return ConstraintInterface
+     */
+    public static function compactConstraint(ConstraintInterface $constraint)
+    {
+        if (!$constraint instanceof MultiConstraint) {
+            return $constraint;
+        }
+
+        $intervals = self::generateIntervals($constraint);
+        $constraints = array();
+        $hasNumericMatchAll = false;
+        $isConjunctive = false;
+
+        $numeric = $intervals['numeric'];
+        $count = \count($numeric);
+        // attempt to convert back 0 - <x + >x - +inf to != x as long as we only have some of those, otherwise bail out of this optimization
+        if ($count > 1 && (string) $numeric[0]->getStart() === (string) Interval::zero() && (string) $numeric[$count-1]->getEnd() === (string) Interval::positiveInfinity()) {
+            $isConjunctive = true;
+            for ($i = 0; $i < $count-1; $i++) {
+                $interval = $numeric[$i];
+                if (isset($numeric[$i+1]) && $interval->getEnd()->getVersion() === $numeric[$i+1]->getStart()->getVersion() && $interval->getEnd()->getOperator() === '<' && $numeric[$i+1]->getStart()->getOperator() === '>') {
+                    $constraints[] = new Constraint('!=', $interval->getEnd()->getVersion());
+                    $numeric[$i+1] = new Interval($interval->getStart(), $numeric[$i+1]->getEnd());
```

Comment:
Why do you overwrite this interval here instead of just doing `$i++`?

Reply:
Yeah this isn't needed anymore now that it checks earlier, just leftover from previous impl.

Will remove this and the other feedback.

Reply:
And as we don't modify anymore, no need to copy the array into $numeric..